peg/toml.peg

```diff
@@ -0,0 +1,273 @@
+#
+# This file is derived from
+#
+# https://github.com/naoina/toml/blob/master/parse.peg ( e401e4e )
+#
+########################################################################
+#
+# Copyright (c) 2014 Naoya Inada <naoina@kuune.org>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+##########################################################################
+#
+# Masatake YAMATO adapted the origina peg file to ctags.
+#
+##########################################################################
+%prefix "ptoml"
+
+%auxil	"struct parserCtx *"
+
+%earlysource {
+    #include "general.h"
+}
+
+%header {
+	struct parserCtx;
+}
+
+%source {
+#include "toml_pre.h"
+}
+
+TOML <- Expression (newline Expression)* newline* !. { ; }
+
+Expression <- (
+   <ws table ws comment? (ws newlineRequired wsnl keyval ws comment?)*> { ; }
+  / ws keyval ws comment?
+  / ws comment?
+  / ws
+)
+
+val <- (
+    <datetime>    { ; }
+  / <float>       { ; }
+  / <integer>     { ; }
+  / <string>      { ; }
+  / <boolean>     { ; }
+  / <array>       { ; }
+  / <inlineTable> { ; }
+)
+
+ws <- [ \t]*
+wsnl <- ([ \t] / newline)*
+
+comment <- '#' (badControl / [\t -\u10FFFF])* { ; }
+
+newline <- ('\r\n' / '\n') { ; }
+
+newlineRequired <- (
+    !newline { ; }
+  / newline
+)
+
+badControl <- <[\u007F\u000B\u000C\u0000-\u0008\u000E-\u001F]> {
+    ;
+}
+
+# -------------------------------------------------------------------------
+# -- Tables
+
+table <- stdTable / arrayTable
+
+stdTable <- '[' ws { tableKeyStart(auxil, false, $0s); } tableKey { tableKeyEnd(auxil); } ws ']'
+
+arrayTable <- '[[' ws { tableKeyStart(auxil, true, $0s); } tableKey { tableKeyEnd(auxil); } ws ']]'
+
+keyval <- { keyvalStart(auxil, $0s); } tableKey { keyvalKeyEnd(auxil); } ws '=' ws val { keyvalValEnd(auxil); }
```

Comment:
Instead of `key`, it seems that we need `tableKey` here for accepting toml-1.0 input .